Create apply utility request 1359

bnlcas opened the pull request:
Added a utility function 'apply.js' to apply a function over the rows/columns of a multidimensional array/matrix. This feature satisfies the requirements of [Request 1359](https://github.com/josdejong/mathjs/issues/1359).

harrysarson commented on src/utils/collection/apply.js:
```diff
@@ -0,0 +1,80 @@
+ * @param {Array | Matrix} mat
+ * @param {number} dim
+ * @param {Function} callback
+ * @return {Array | Matrix} res
```
Would it be possible to give fuller descriptions of these parameters?

bnlcas replied:
Yes, will do.

harrysarson commented on src/utils/collection/apply.js:
```diff
@@ -0,0 +1,80 @@
+ * @param {Function} callback
+ * @return {Array | Matrix} res
+ */
+module.exports = function (mat, dim, callback) {
```
Would this be better implemented as a typed function?

bnlcas replied:
Potentially. I was following the form of the other utility functions like reduce.js or deepMap.js which are not implemented as typed functions. I would be happy to make this change if you think it would be improve the quality of the function.

harrysarson commented:
The main question I have is why does the implementation of apply involve transposing the matrix? Forgive me if this is an obvious question.

bnlcas commented:
@harrysarson The use of transpose is a little esoteric. The reason that apply transposes the matrix is so that the callback function can be applied sequentially to a list of arrays. This allows the callback function to be applied along an arbitrary axis of an n-dimensional array (things get complicated for n > 3). The logic of this function was borrowed from the current implementation of reduce.js.

josdejong commented:
Thanks @bnlcas, I like this function already :)

Some feedback points:

1. You've created `apply` under utils functions. The functions under `/src/utils`, which is used for internal util functions that are used from multiple places in the code base. The public matrix related functions are located in `/src/function/matrix`, can you move `apply` there at add it to the `index.js` file in `src/function/matrix`?
2. All public functions in mathjs are implemented as a `typed-function` which makes sure they automatically throw neat type errors when called with invalid inputs. You can for example look at `concat.js` for inspiration.
3. Just thinking aloud: would it make sense to support `apply(matrix, sum)` (no dimension specified) to simply calculate the sum of all values?
4. I think we also need a _transform_ for this function: the expression parser uses one-based indexes, and we have transform functions to change indexes into zero-based ones. See for example `src/expression/transform/concat.transform.js` and `src/expression/transform/index.transform.js`.

harrysarson commented:
> The logic of this function was borrowed from the current implementation of reduce.js.

If this is the approach we already use then it makes sense to me. 👍

bnlcas commented:
@josdejong, in response to your comments:
1.) I am happy with this being a public function - I will move it.
2.) Agreed that this should be a typed-function if public.
3.) I think that apply(matrix, sum) returning the sum of everything feels a little strange to me. This could be interpreted either as being 'sum(flatten(X))', or as 'apply(apply(X, sum,0), sum, 0)'. These are the same for a linear function like sum, but not in the case of std.
4.) I will add an expression function for this.
5.) In addition to the expression function, should there be a documentation for this since it will now be public?

harrysarson commented:
> 3.) I think that apply(matrix, sum) returning the sum of everything feels a little strange to me. This could be interpreted either as being 'sum(flatten(X))', or as 'apply(apply(X, sum,0), sum, 0)'. These are the same for a linear function like sum, but not in the case of std.

Less "magic" default arguments that have potentially surprising behaviour is better in my book so I would vote against `apply(matrix, sum)` returning the sum of everything.

> 5.) In addition to the expression function, should there be a documentation for this since it will now be public?

Documentation is automatically generated based on jsdoc comments. The comments in `src/function/matrix/apply.js` look like they are almost there to me. With some examples and related function references I think it would be good to go :)

If you look at the source of some other files in the `src/function/matrix/` it should give you an idea of how the comments usually look.

I don't know why the tests failed - it looked unrelated. Will try re-running them.

josdejong commented:
Good arguments, thanks, let's not go for `apply(matrix, sum)` but keep things simple and straight forward :+1: 

The unit tests failed because decimal.js@10.1.0 not exists (#1421). Not sure whether greenkeeper made a mistake or that this version was immediately unpublished and succeeded by decimal.js@10.1.1. 

Tests are still failing but now for something that is actually related to the changes made, the requires in `apply.js` don't point correctly to the actual files:
```js
const arraySize = require('../array').size
const isMatrix = require('../collection/isMatrix')
```

harrysarson commented on src/function/matrix/apply.js:
```diff
@@ -0,0 +1,82 @@
+'use strict'
+
+const arraySize = require('../array').size
```
I think this line is causing the CI error

josdejong commented:
Indeed. Functionally everything is in place. What is left over I think is:

2.) Turn `apply` into a typed-function.
4.) Creating a transform function for it.
5.) Write out docs in the comment section of the function (look at other functions for an example)

@bnlcas do you have enough pointers to continue?

bnlcas commented:
@josdejong I just made changes to turn apply into a typed function, and I updated the unit test to reflect this change. Additionally, I updated the comments to bring them more inline with the documentation standards. Tomorrow I will work making this function work with expressions. I'll touch back later this week.

josdejong commented:
Looks good @bnlcas !

@harrysarson how do you think about the codecov plugin? It's nice but also maybe a bit too aggressive. Adding unit tests just to get the coverage higher isn't a helpful goal in itself.

harrysarson commented:
Yeah, I think it is very useful to have codecov track our coverage over time.

However, the red ticks are not good

bnlcas commented:
In general tracking code coverage is a good idea. I think that is happening here is that 'apply.transform.js' does not have a test since it is essentially a wrapper for 'apply.js'.

Because many new features will include require the addition of 'feature.transform.js', this problem will keep happening in the future. We could try tweaking the thresholds, but this will not necessarily solve the problem (small commits for example could fall under the new threshold).
I think it would be best to either create tests for the '*.transform.js' functions, or to exclude this section of the code base from the test coverage requirements.

josdejong commented:
The problem (for this PR) may be solved already when adding a unit test for the transform function, that shouldn't be much work. Can you do that @bnlcas ? Is there anything else left you want to improve in this PR? All looks good to me.

bnlcas commented:
@josdejong I just added a test for 'apply.transform.js' - this feels like the right approach. Beyond that I think that things are in a good place here.

josdejong commented:
Thanks @bnlcas, this new function is a really nice addition to mathjs! Merging your work now, will do a deployment in a couple of days.

josdejong commented:
@bnlcas `apply` is now available in `v5.8.0`, thanks again!
